### DISCUSSION TITLE
Use TypeScript types for complex queries across codebase

### DISCUSSION THREAD

**Discussion by stefan-girlich** [supabase/supabase#21838] [Questions]
Creating types from a query (see [docs](https://supabase.com/docs/reference/javascript/select#response-types-for-complex-queries)) works fine for me, but I don't see an obvious way of using the inferred type elsewhere in my codebase, e.g. in a React component defined in its own file.

### 1. Use PostgreSQL views instead
far less convenient and doesn't allow for nested output data

### 2. Manually transform the result data to a new statically defined type
less convenient and would need to happen in the same function where the inferred type lives or pass inferred type via generics

### 3. Infer the type from a function
works, but requires extra effort:
```ts
type AsyncReturnType<T extends (...args: any) => Promise<any>> = T extends (
  ...args: any
) => Promise<infer R>
  ? R
  : any;

const fetchBatches = async (supabase: SupabaseClient<Database>) => {
  const batchesWithEventsQuery = supabase.from("batches").select(
    `*,
     batch_events(
       id, type, batch_event_types(id, name)
     )`
  );
  type BatchesWithEventsInferred = QueryData<typeof batchesWithEventsQuery>;
  const { data } = await batchesWithEventsQuery;
  return data as BatchesWithEventsInferred;
};

type BatchesWithEvents = AsyncReturnType<typeof fetchBatches>;
export type BatchWithEvents = BatchesWithEvents[number];
```

What is the most practical way to use TypeScript types for queries?

**Comment by yamz8** [comment]
any updates about that?

**Reply by stefan-girlich** [reply]
Unfortunately no new learnings from my end; I stuck with the option #3 for now.

**Reply by gabriel-peter** [reply]
Thanks for this helper! Not ideal, but it works for my use-cases well enough. :)

**Comment by louisprp** [comment]
This has worked great for me, although with a slightly different format.

Having a function defined like this

```
export const getOrders = async (
  client: SupabaseClient<Database>,
  signal?: AbortSignal
) => {
  const query = client
    .from("orders")
    .select(
      "id, customer_id:users!orders_customer_id_fkey(*), supplier_id:users!orders_supplier_id_fkey(*), status, created_at, orders_products (*)"
    )
  if (signal) {
    query.abortSignal(signal)
  }
  return query
}
```

I can now get the type of a single order like this:

`export type Order = NonNullable<AsyncReturnType<typeof getOrders>["data"]>[number]`

Where `AsyncReturnType` looks like this:

```
// Types for getting the return type of an asynchronous function
type PromiseType<T> = T extends Promise<infer U> ? U : never

export type ErrorType = { error: string }

/**
 * Represents the return type of an asynchronous function.
 * It extracts the resolved value from a Promise and excludes any potential error type.
 * Used primarily to get supabase action return types.
 */
export type AsyncReturnType<T extends (...args: any) => Promise<any>> = Exclude<
  PromiseType<ReturnType<T>>,
  ErrorType
>
```

**Reply by Ltwoz** [reply]
Thanks for this.

**Reply by SkipDead** [reply]
you are a legend. Thanks !

**Comment by Thimotio17** [comment]
A practical approach is to keep the query builder in a shared data-access module and derive the exported type directly from it using `QueryData`.

```ts
import {
  QueryData,
  SupabaseClient,
} from '@supabase/supabase-js'
import type { Database } from './database.types'

export const batchesWithEventsQuery = (
  supabase: SupabaseClient<Database>
) =>
  supabase
    .from('batches')
    .select(`
      *,
      batch_events (
        id,
        type,
        batch_event_types (
          id,
          name
        )
      )
    `)

export type BatchesWithEvents = QueryData<
  ReturnType<typeof batchesWithEventsQuery>
>

export type BatchWithEvents = BatchesWithEvents[number]
